src/main/java/uk/gov/justice/hmiprobation/casesampler/services/AllocationCalculator.kt

```diff
@@ -0,0 +1,66 @@
+package uk.gov.justice.hmiprobation.casesampler.services
+
+import org.slf4j.LoggerFactory
+import uk.gov.justice.hmiprobation.casesampler.dto.Case
+import uk.gov.justice.hmiprobation.casesampler.utils.Size
+import uk.gov.justice.hmiprobation.casesampler.utils.Size.SampleSize
+import uk.gov.justice.hmiprobation.casesampler.utils.calculateSampleSize
+import kotlin.random.Random
+
+data class Info(val id: String, val size: Size)
+
+data class AllocationData(val cluster: Info, val ldu: Info, val ro: Info)
+
+data class RoAllocation(val allocationData: AllocationData, val cases: List<Case>) {
+
+    fun getRandomSamples(random: Random = Random) = cases.shuffled(random).take(allocationData.ro.size.count)
+}
+
+/**
+ *  Calculates the proportions of samples that are required for each RO whilst maintaining proportionality across
+ *  - Cluster
+ *  - Local Delivery Unit (LDU)
+ *  - Responsible Officer (RO) up to maximum of n
+ */
+class AllocationCalculator(val roAllocationAdjuster: RoAllocationAdjuster) {
+
+    fun calculate(size: SampleSize, cases: List<Case>): List<RoAllocation> {
+        val casesByCluster = cases.groupBy { it.cluster }
+
+        val clusterSizes = calculateSampleSize(size.numberOfSamples, casesByCluster)
+
+        return clusterSizes.fold(mutableListOf()) { result, (cluster, size) ->
+            result.addAll(toRoAllocation(Info(cluster, size), casesByCluster[cluster]!!))
```

Comment:
I wonder why it thinks `casesByCluster[cluster]` can be null. It would be nice to find away around it so you don't need `!!`

Reply:
I was tempted with a getOrDefault returning an empty list if not present. Thought it might be more explicit to say `!!` as it should always be there